starlarktime/time.go

```diff
@@ -0,0 +1,507 @@
+		"hour",
+		"minute",
+		"second",
+		"nanosecond",
+		"unix",
+		"unix_nano",
+	)
+}
+
+// CompareSameType implements comparison of two Time values. required by
+// starlark.Comparable interface.
+func (t Time) CompareSameType(op syntax.Token, yV starlark.Value, depth int) (bool, error) {
+	x := time.Time(t)
+	y := time.Time(yV.(Time))
+	cp := 0
+	if x.Before(y) {
+		cp = -1
+	} else if x.After(y) {
+		cp = 1
+	}
+	return threeway(op, cp), nil
+}
+
+// Binary implements binary operators, which satisfies the starlark.HasBinary
+// interface
+//    time + duration = time
+//    time - duration = time
+//    time - time = duration
+func (t Time) Binary(op syntax.Token, yV starlark.Value, side starlark.Side) (starlark.Value, error) {
+	x := time.Time(t)
+
+	switch op {
+	case syntax.PLUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Time(x.Add(time.Duration(y))), nil
+		}
+	case syntax.MINUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Time(x.Add(time.Duration(-y))), nil
+		case Time:
+			// time - time = duration
+			if side == starlark.Left {
+				return Duration(x.Sub(time.Time(y))), nil
+			} else {
+				return Duration(time.Time(y).Sub(x)), nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+var timeMethods = map[string]builtinMethod{
+	"in_location": timeIn,
+	"format":      timeFormat,
+}
+
+func timeFormat(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x string
+	if err := starlark.UnpackPositionalArgs("format", args, kwargs, 1, &x); err != nil {
+		return nil, err
+	}
+
+	recv := time.Time(recV.(Time))
+	return starlark.String(recv.Format(x)), nil
+}
+
+func timeIn(fnname string, recV starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x string
+	if err := starlark.UnpackPositionalArgs("in_location", args, kwargs, 1, &x); err != nil {
+		return nil, err
+	}
+	loc, err := time.LoadLocation(x)
+	if err != nil {
+		return nil, err
+	}
+
+	recv := time.Time(recV.(Time))
+	return Time(recv.In(loc)), nil
+}
+
+type builtinMethod func(fnname string, recv starlark.Value, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error)
+
+func builtinAttr(recv starlark.Value, name string, methods map[string]builtinMethod) (starlark.Value, error) {
+	method := methods[name]
+	if method == nil {
+		return nil, nil // no such method
+	}
+
+	// Allocate a closure over 'method'.
+	impl := func(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+		return method(b.Name(), b.Receiver(), args, kwargs)
+	}
+	return starlark.NewBuiltin(name, impl).BindReceiver(recv), nil
+}
+
+func builtinAttrNames(methods map[string]builtinMethod) []string {
+	names := make([]string, 0, len(methods))
+	for name := range methods {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+// threeway interprets a three-way comparison value cmp (-1, 0, +1)
```

Comment:
FYI: in hindsight, the API design that required `threeway` was a mistake. Originally it was intended to support floating point NaNs according to IEEE 754, which requires that x < NaN, Nan < x and NaN == x are all false, so you can't represent the result of a "float op float" comparison using a three-valued result (smaller, bigger, equal): one must also know the query operator. But we changed things recently so that floats are totally ordered (NaN > +Inf), otherwise you can't sort floats, so CompareSameType could in principle be simplified to:

```
CompareSameType(v Value, depth int) (signum int, err error)
```